Sources/GRPCCore/GRPCClient.swift

```diff
@@ -0,0 +1,823 @@
+/// client.methodConfigurationOverrides = registry
+/// ```
+///
+/// ## Starting and stopping the client
+///
+/// Once you have configured the client, call ``run()`` to start it. Calling ``run()`` connects to the given
+/// transport.
+///
+/// ```swift
+/// // Start running the client.
+/// try await client.run()
+/// ```
+///
+/// The ``run()`` method won't return until the client has finished handling all requests. You can
+/// signal to the client that it should stop creating new request streams by calling ``close()``.
+/// This gives the client enough time to drain any requests already in flight. To stop the client more abruptly
+/// you can cancel the task running your client. If your application requires additional resources
+/// that need their lifecycles managed you should consider using [Swift Service
+/// Lifecycle](https://github.com/swift-server/swift-service-lifecycle).
+@available(macOS 13.0, iOS 16.0, watchOS 9.0, tvOS 16.0, *)
+public final class GRPCClient: Sendable {
+  /// A collection of ``ClientInterceptor`` implementations which are applied to all accepted
+  /// RPCs.
+  ///
+  /// RPCs are intercepted in the order that interceptors are added. That is, a request sent from the client to
+  /// the server will first be intercepted by the first added interceptor followed by the second, and so on.
+  /// For responses from the server, they'll be applied in the opposite order.
+  public var interceptors: Interceptors {
+    get {
+      self.storage.withLockedValue { $0.interceptors }
+    }
+    set {
+      self.storage.withLockedValue { storage in
+        if case .notStarted = storage.state {
+          storage.interceptors = newValue
+        }
+      }
+    }
+  }
+
+  /// A ``MethodConfigurationRegistry`` containing ``MethodConfiguration``s for calls
+  /// made from this ``Client``.
+  ///
+  /// - Note: These configurations will override those configurations set in the ``ClientTransport``.
+  public var methodConfigurationOverrides: MethodConfigurationRegistry {
+    get {
+      self.storage.withLockedValue { $0.methodConfigurationOverrides }
+    }
+    set {
+      self.storage.withLockedValue { storage in
+        if case .notStarted = storage.state {
+          storage.methodConfigurationOverrides = newValue
+        }
+      }
+    }
+  }
+
+  /// The state of the client.
+  private enum State {
+    /// The client hasn't been started yet. Can transition to `running` or `stopped`.
+    case notStarted
+    /// The client is running and can send RPCs. Can transition to `stopping`.
+    case running
+    /// The client is stopping and no new RPCs will be sent. Existing RPCs may run to
+    /// completion. May transition to `stopped`.
+    case stopping
+    /// The client has stopped, no RPCs are in flight and no more will be accepted. This state
+    /// is terminal.
+    case stopped
+  }
+
+  /// Underlying storage for the client.
+  private struct Storage {
```

Comment:
This is the opposite of what we normally do. We put the storage into the associated values of the `State` enum. The advantage of that is that once we transition to `stopped` we free all of the retained storage.

Reply:
Not always. We tend to do that when the state management and dependencies are complicated. That's not the case here.

Reply:
Sure but I still think there is benefits to that approach like freeing up the resources as soon as possible. Right now we are going to retain the interceptors for longer than we need.